### PR TITLE
`Programming exercises`: Fix missing gradle feedback on Jenkins

### LINKED ISSUE #6434: `Programming Exercises`: Gradle test case feedback cut in half on Jenkins setups
### Describe the bug

In Jenkins setups the test case feedbacks for Gradle exercises is cut in half.

This originates here:
https://github.com/ls1intum/Artemis/blob/a9963ac2b07643b65808da9f0c1446f9d8917f57/src/main/java/de/tum/in/www1/artemis/repository/FeedbackRepository.java#L212

The duplication might be true for Bamboo.

### To Reproduce

Artemis with Jenkins setup required.

1. Create a new Gradle+Java exercise.
2. Add a test case like
   ```java
    @Test
    void longFeedback() {
        fail("""
            a
            b
            c
            d
            """);
    }
    ```
3. Check the solution build result. The feedback is shown as in the screenshot below.

### Expected behavior

The full feedback is shown.

### Screenshots

image

### Which version of Artemis are you seeing the problem on?

6.1.2

### What browsers are you seeing the problem on?

Firefox

### Additional context

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by Strohgelaender**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 11 (LocalVC).

### Motivation and Context
Fixes #6434

Bamboo duplicates Gradle Feedback and therefore Artemis cuts it in half. However, this is not the case for Jenkis setup, here Feedback currently gets lost.

### Description
I first tried to cut the feedback directly in the bamboo plugin, but I was not able to detect the exercise type (Maven vs Gradle) there. Now we check if we are on a Bamboo setup and only cut the feedback if this is the case.

In order to use the `ProfileService`, I move the code from a static method in `ResultRepository` into its own `ProgrammingExerciseFeedbackCreationService`.  This is where most of the changed lines in this PR come from, the only real change is in the additional check.

### Steps for Testing
Do the following on all 3 supported CI setups (Bamboo, Jenkins, LocalCI):

1. Create a Gradle PE with a test case that produces multiline feedback, e.g.:

```
@Test
 void longFeedback() {
     fail("""
         a
         b
         c
         d
         """);
 }
```

2. Submit a solution as a student
3. Check that the long feedback is shown correctly.

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Reviv
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### Test Coverage
changes covered by additional tests.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackCreationService.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+            String generalTimeOutExceptionText = matcher.group(1);
+            return timeoutDetailText + "\n" + exceptionPrefix + generalTimeOutExceptionText.trim();
+        }
+
+        // Filter out unneeded Exception classnames
+        if (programmingLanguage == ProgrammingLanguage.JAVA || programmingLanguage == ProgrammingLanguage.KOTLIN) {
+            var messageWithoutStackTrace = message.lines().takeWhile(IS_NOT_STACK_TRACE_LINE).collect(Collectors.joining("\n")).trim();
+
+            // the feedback from gradle test result is duplicated on bamboo therefore it's cut in half
+            if (projectType != null && projectType.isGradle() && profileService.isBamboo()) {
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackCreationService.java:79]: This line is the main change of this PR: Only cut the feedback on a bamboo setup.

**Review by b-fein — APPROVED** [review]
Thanks! Tested on TS2, code changes also look good. :+1:

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackCreationService.java (hunk 1 above)
Comment (jakubriegel) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackCreationService.java:79]: Is Bamboo it always cutting them in half or does it have some maximum length?

**Review by jakubriegel — APPROVED** [review]

**Review by jakubriegel — DISMISSED** [review]
Code looks good 👍🏻 

Works fine on Bamboo

File: src/test/java/de/tum/in/www1/artemis/hestia/CodeHintIntegrationTest.java [hunk 2]
```diff
@@ -89,7 +92,9 @@ private void addCodeHints() {
         programmingExerciseUtilService.addTasksToProgrammingExercise(exercise);
         programmingExerciseUtilService.addSolutionEntriesToProgrammingExercise(exercise);
         programmingExerciseUtilService.addCodeHintsToProgrammingExercise(exercise);
-        codeHint = codeHintRepository.findByIdWithSolutionEntriesElseThrow(codeHintRepository.findByExerciseId(exercise.getId()).stream().findAny().orElseThrow().getId());
+        Set<CodeHint> hints = codeHintRepository.findByExerciseId(exercise.getId());
+        codeHint = hints.stream().filter(hint -> hint.getProgrammingExerciseTask().getTaskName().equals("Task for test1")).findFirst().orElseThrow();
```
Comment (MaximilianAnzinger) [inline src/test/java/de/tum/in/www1/artemis/hestia/CodeHintIntegrationTest.java:96]:
```suggestion
        codeHint = hints.stream().filter(hint -> "Task for test1".equals()).findFirst().orElseThrow(hint.getProgrammingExerciseTask().getTaskName());
```
Teamscale warning

File: src/test/java/de/tum/in/www1/artemis/hestia/CodeHintIntegrationTest.java [hunk 3]
```diff
@@ -89,7 +92,9 @@ private void addCodeHints() {
         programmingExerciseUtilService.addTasksToProgrammingExercise(exercise);
         programmingExerciseUtilService.addSolutionEntriesToProgrammingExercise(exercise);
         programmingExerciseUtilService.addCodeHintsToProgrammingExercise(exercise);
-        codeHint = codeHintRepository.findByIdWithSolutionEntriesElseThrow(codeHintRepository.findByExerciseId(exercise.getId()).stream().findAny().orElseThrow().getId());
+        Set<CodeHint> hints = codeHintRepository.findByExerciseId(exercise.getId());
+        codeHint = hints.stream().filter(hint -> "Task for test1".equals()).findFirst().orElseThrow(hint.getProgrammingExerciseTask().getTaskName());
```
Comment (b-fein) [inline src/test/java/de/tum/in/www1/artemis/hestia/CodeHintIntegrationTest.java:96]:
```suggestion
        codeHint = hints.stream().filter(hint -> "Task for test1".equals(hint.getProgrammingExerciseTask().getTaskName())).findFirst().orElseThrow();
```

Comment (Strohgelaender) [reply src/test/java/de/tum/in/www1/artemis/hestia/CodeHintIntegrationTest.java]: Thanks!

**Review by jakubriegel — APPROVED** [review]
Tested on testing session

**Review by dearjasmina — APPROVED** [review]
tested in a testing session, works good (ts1 and ts2)

**Review by b-fein — APPROVED** [review]
code re-approve

**Review by Kroko-fant — APPROVED** [review]
tested in the testing session yesterday and it works fine

**Pull request merged**